src/core/task.ts

```diff
@@ -82,13 +80,46 @@ export async function pause(method: Metric): Promise<void> {
     // Instead, we will turn async task into a sync task and maximize our chances of getting some data back.
     if (method in tracker) {
         stop(method);
-        await wait();
+        tracker[method].yield = (await wait()).timeRemaining();
         resume(method);
     }
 }
 
-async function wait(): Promise<number> {
-    return new Promise<number>((resolve: FrameRequestCallback): void => {
-        requestAnimationFrame(resolve);
+async function wait(): Promise<RequestIdleCallbackDeadline> {
+    return new Promise<RequestIdleCallbackDeadline>((resolve: (deadline: RequestIdleCallbackDeadline) => void): void => {
+        requestIdleCallback(resolve, { timeout: idleTimeout });
     });
 }
+
+// Use native implementation of requestIdleCallback if it exists.
+// Otherwise, fall back to a custom implementation using requestAnimationFrame & MessageChannel.
+// While it's not possible to build a perfect polyfill given the nature of this API, the following code attempts to get close.
+// Background context: requestAnimationFrame invokes the js code right before: style, layout and paint computation within the frame.
+// This means, that any code that runs as part of requestAnimationFrame will by default be blocking in nature. Not what we want.
+// For non-blocking behavior, We need to know when browser has finished painiting. This can be accomplished in two different ways (hacks):
+//   (1) Use MessageChannel to pass the message, and browser will receive the message right after pain event has occured.
+//   (2) Use setTimeout call within requestAnimationFrame. This also works, but there's a risk that browser may throttle setTimeout calls.
+// Given this information, we are currently using (1) from above. More information on (2) as well as some additional context is below:
+// https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Performance_best_practices_for_Firefox_fe_engineers
+let requestIdleCallback = window["requestIdleCallback"] ||
+function(callback: (deadline: RequestIdleCallbackDeadline) => void, options: RequestIdleCallbackOptions): void {
```

Comment:
Super nit: Could you make this a top-level function so that it is testable and then do window["...."] || myfunction

Reply:
Done